### ISSUE TITLE
Add cookie tags

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#1462]
### What is the feature?

Since we've been making changes towards being "the privacy CMS" does it make sense to take this a step further and either:

* Bake in cookie opt-in/out tags into core, and/or;
* Add a conditional tag for checking cookie status in a page request.

There have been various attempts at plugins or solutions over the years:

* https://github.com/jools-r/jquery-cookiebar
* https://github.com/cara-tm/pat_eu_cookies_law
* https://github.com/NicolasGraph/oui_cookie
* https://textpattern.tips/setting-cookies-for-eu-legislation

Is there anything elegant from all those we can capture and distill into a lowest-common-denominator approach to satisfy this feature?

Just thinking it might be handy at a minimum to permit something like:

~~~ html
<txp:if_cookie name="opt_in" value="true">
   // Render your evil tracking scripts here
</txp:if_cookie>
~~~

Further, if we could build a container tag that allowed people to add/customise a bar, with appropriate 'accept' or 'decline' buttons then maybe that'd be useful too. Links to any privacy policy or script to gather the in/out status of various cookies is probably outside the scope.

Not sure if any of that's possible given the infinite number of ways people might want to present the options.

Maybe it could pass back to the server (optionally via Ajax?) the result to set the given cookie values - whether they be a set of opt-in or opt-out values. Not entirely sure how that would work.

I'd be happy with just a conditional cookie checking tag, but if there's any scope to build a simple, extensible, cookie tracking solution in core too, then perhaps we could look at it. Alternatively, we might decide this entire idea is not worthy of a core solution and leave it to plugins.

Thoughts?

**Comment by petecooper** [comment]
For me, the cookie opt-in/-out bar is plugin territory. I don't have a strong feeling on this, so please take this with a suitable amount of salt.

There's GDPR, there's now CCPA, and it's a fairly safe bet that more will come. I am not a lawyer, and the two I know are well-paid for a reason. Organisations who use all the tracking stuff will, in an ideal world, have people who know the legalities of it. They can use Textpattern as the framework, and their own know-how (and designers, and legal counsel) for the presentation, tracking, and so forth.

Anything we offer in core for this stuff may make us liable, even in some convoluted way, and that does worry me.

**Comment by Bloke** [comment]
Good point about the cookie bar. Hadn't thought of that angle.

That's fine by me. We can just go with a conditional tag to test a cookie status if people think that's worth pursuing. Much easier to implement too. Like, 5-10 lines of code or something.

**Comment by petecooper** [comment]
>We can just go with a conditional tag to test a cookie status if people think that's worth pursuing. Much easier to implement too.

File under '4.9' or 'distract @petecooper and sneak it by him for 4.8'?

**Comment by Bloke** [comment]
Ooooh, [is that Halley's comet](https://github.com/textpattern/textpattern/commit/03b334a3356b5deadbf87fa22ae7e5c0ad1c02e2) ?

**Comment by petecooper** [comment]
NOW LISTEN HERE YOU LITTLE SH–

**Comment by philwareham** [comment]
Needs documentation, obviously Do you want me to create a placeholder page in the docs site for you to fill out as and when :)

**Comment by Bloke** [comment]
I'll add it, no worries. Was just gonna wait until we thought it was worth keeping.

**Comment by bloatware** [comment]
We could also rename it `if_request` end extend to `get/post` types?

**Comment by Bloke** [comment]
@bloatware I did wonder about that, but figured it was already covered with `<txp:page_url>`. Maybe not. If you think that would be useful and a quick win, yes, do it.

Maybe we should extend `<txp:page_url>` to check for cookies instead of introducing a new tag? Or is that too esoteric? Hmmm, but it's not container/conditional so maybe that won't work.

I'm fine with whatever. If it makes sense to rename the tag and extend its capabilities, that's perfect for me. With global escape, people can be covered from an injection standpoint when they use it to read URL vars.

**Comment by bloatware** [comment]
Yes, `<txp:page_url />` is not conditional (though container now) and overcharged, so there would be no overlap. And in `<txp:if_request />` we would only check the value, not output it, so it's ok injection-wise (which is possible via cookies anyway). But then we would probably need a paired `<txp:request />` tag too... look Pete, two moons!

**Comment by petecooper** [comment]
Sigh.

**Comment by petecooper** [comment]
How viable is a `<txp:request />` and `<txp:if_request />` conditional, given the current workload and scope?

We can defer until `beta.3` if @colak's issues in https://github.com/textpattern/textpattern/issues/1463 are diagnosed and resolved.

(And yes, I'll look the other way.)

**Comment by Bloke** [comment]
`<txp:if_request/>` just needs a `type` attribute adding to switch between (or cycle through) particular flavours of input: cookie, post or get. Presumably, we could get it to default to `cookie, post, get` which would check those sources in that order and as soon as it found a match, would bomb out with a yes, otherwise no. Bonus marks for setting a context so that any embedded `<txp:request>` was auto-populated with the matching content.

Not sure exactly how `<txp:request>` would interplay with the conditional. Oleg presumably has ideas so I'd leave this in his capable hands unless he has no time to do it today. In which case, feel free to describe them and I'll do it.

Otherwise, I could back this tag out for now and defer to beta.3/rc.1 if colak's issue is deemed resolved before we add the tags.

**Comment by petecooper** [comment]
>Otherwise, I could back this tag out for now and defer to beta.3/rc.1 if colak's issue is deemed resolved in short order.

This sounds like a great plan.

**Comment by Bloke** [comment]
Okay, I've [backed it out](https://github.com/textpattern/textpattern/commit/8bc48da3d35a7176b84783891b874c9fbfa43820) for now. Let's sit on this and think about it a little more.

**Comment by bloatware** [comment]
Too [late](https://github.com/textpattern/textpattern/commit/0108cf1de23e22d5fd737b67409de23deaebcaca) for `if_request`, but let's leave `request` (and `if_request` tweaks) for b3.

Extra bonus: `if_variable` has now `match` and `separator` attributes too.

**Comment by philwareham** [comment]
OK, let's do beta2 :) I've just put the latest translations I had in, so all ready from my side.

**Comment by Bloke** [comment]
Nice work, Oleg. Better than my implementation would have been. And funnily enough, the other day I tried to use a pattern match for `<txp:if_variable>` and only after I'd written it did I remember that the attribute was confined to `<txp:if_custom_field>`. Centralizing the txp_match concept gives us some great scope to introduce pattern matching in other conditionals that make sense later.

**Comment by Bloke** [comment]
Okay, let's freeze commits and I'll bundle up a release this evening.

**Comment by bloatware** [comment]
@Bloke should we add `_SERVER` to the list? I saw you mention it in history, and it can be useful to check referer and all.

**Comment by Bloke** [comment]
I meant server variables in general, of which REQUEST are a subset, but yeah, for the sake of twenty characters or something, sure, the tag can support those, and they are handy as you say. Feel free to update HISTORY as well please, if necessary.

I'm just compiling the blog post from the HISTORY changes since beta.1 so you've got a little while before I create the release.

**Comment by bloatware** [comment]
Ok, server added, but not sessions because they need to be started first.

**Comment by Bloke** [comment]
Oh yeah, good spot. Nice one, thanks.

**Comment by petecooper** [comment]
@Bloke - based on your OP and subsequent commits, are we broadly OK to mark this closed, or am I missing something? Thank you!

**Comment by Bloke** [comment]
Yep. We have the 'if' cookie. Do we need the 'cookie value' tag as companion, or can that wait?

**Comment by petecooper** [comment]
Great! We can defer cookie value to a point release, I think - this is a solid step forward. Will close.
